pkg/runner/mounting_runner.go

```diff
@@ -0,0 +1,75 @@
+package runner
+
+import (
+	"context"
+	"path/filepath"
+
+	"github.com/buildbarn/bb-remote-execution/pkg/proto/configuration/bb_runner"
+	runner_pb "github.com/buildbarn/bb-remote-execution/pkg/proto/runner"
+	"github.com/buildbarn/bb-storage/pkg/filesystem"
+	"github.com/buildbarn/bb-storage/pkg/filesystem/path"
+	"github.com/buildbarn/bb-storage/pkg/util"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+type mountingRunner struct {
+	base           runner_pb.RunnerServer
+	buildDirectory filesystem.Directory
+	mount          *bb_runner.InputMountOptions
+}
+
+// NewMountingRunner is a decorator for Runner
+// that mounts `mount` before running a build action.
+//
+// This decorator can be used for chroot runners
+// that must mount special filesystems into the input root.
+func NewMountingRunner(base runner_pb.RunnerServer, buildDirectory filesystem.Directory, mount *bb_runner.InputMountOptions) runner_pb.RunnerServer {
+	return &mountingRunner{
+		buildDirectory: buildDirectory,
+		mount:          mount,
+		base:           base,
+	}
+}
+
+func (r *mountingRunner) Run(ctx context.Context, request *runner_pb.RunRequest) (response *runner_pb.RunResponse, err error) {
+	pathResolver := buildDirectoryPathResolver{
+		stack: util.NewNonEmptyStack(filesystem.NopDirectoryCloser(r.buildDirectory)),
+	}
+	scopeWalker := path.NewRelativeScopeWalker(&pathResolver)
+	defer pathResolver.closeAll()
+
+	fullpath := filepath.Join(request.InputRootDirectory, r.mount.Mountpoint)
```

Comment:
My suggestion would be to not use `filepath.Join()` here. It assumes that the native path separator is used, even though `request.InputRootDirectory` always uses forward slashes, even on Windows. There is also no need to do that if you just call `path.Resolve()` multiple times:
```go
if err := path.Resolve(request.InputRootDirectory, path.NewRelativeScopeWalker(&pathResolver)); err != nil {
    ...
}
if err := path.Resolve(r.mount.Mountpoint, path.NewRelativeScopeWalker(&pathResolver)); err != nil {
    ...
}
```

I would also strongly advise against using the same PathResolver for both processing the input root directory and the mount point. Notice how the path resolver contains a stack. That stack is there to process `".."` entries. Because the stack isn't cleared/truncated, a mount point containing `".."` components might leave the input root directory.

This is not a big issue right now, because the mount point is fully specified in config. But if someone ever comes along and wants the mount point path resolution to respect symlinks, it can become dangerous fairly quickly.